flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java

```diff
@@ -87,12 +87,13 @@ public BlobClient(InetSocketAddress serverAddress, Configuration clientConfig) t
 				LOG.info("Using ssl connection to the blob server");
 
 				socket = SSLUtils.createSSLClientSocketFactory(clientConfig).createSocket(
-					serverAddress.getAddress(),
+					serverAddress.getHostName(),
 					serverAddress.getPort());
 			}
 			else {
-				socket = new Socket();
-				socket.connect(serverAddress);
+				// Establish the socket using the hostname and port. This avoids a potential issue where the
+				// InetSocketAddress can cache a failure in hostname resolution forever.
+				socket = new Socket(serverAddress.getHostName(), serverAddress.getPort());
```

Comment:
How about doing the lookup only if the `InetSocketAddress` was not resolved?

```
socket = new Socket();
if (serverAddress.isUnresolved()) {
	socket.connect(new InetSocketAddress(serverAddress.getHostName(), serverAddress.getPort()));
} else {
	socket.connect(serverAddress);
}
```

Reply:
We can do that, but I'm not sure it's the best idea. Since the resolution in `InetSocketAddress` only happens when the object is instantiated, you could cache an old name resolution. The most likely reason for this to happen would be a deployment environment such as Kubernetes. If the JobManger pod is restarted, then a new pod will take over and will get assigned a new IP address. Kubernetes will automatically update the DNS entry using KubeDNS for the service name. I think that means you could have a situation where the `InetSocketAdress` says it's resolved, but it cached an old resolution. The safest thing is to force the DNS resolution when we make the connection. Both Java and the kernel will already cache the DNS entry based on the TTL in the DSN record, so I don't think we gain much by adding an additional caching layer in the `InetSocketAddress` itself.

Let me know if you agree or if we should still consider the change.

Reply:
Hi @joey , I've replied in the JIRA as I think this is a more fundamental question about the root cause of what we are trying to solve here.